pkg/controller/certificates/approver/sarapprove.go

```diff
@@ -169,3 +174,20 @@ func usagesToSet(usages []capi.KeyUsage) sets.String {
 	}
 	return result
 }
+
+func isSelfNodeServingCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if csr.Spec.Username != x509cr.Subject.CommonName {
+		return false
+	}
+	if csr.Spec.SignerName != capi.KubeletServingSignerName {
+		return false
+	}
+	// dnsNames and ipAddresses are not validated because this generic controller cannot check infra details.
+	// letting any kubelet get a serving cert for any DNS/IP still seems better than turning off all TLS verification
+	// (assuming that the signer uses a distinct CA for this purpose and therefore nodes cannot MITM unrelated traffic).
+	// combined with --kubelet-validate-node-name, we can know that TLS connections from API server to kubelet
+	// are landing on a node that at one point could authenticate to the API server as that node name.  Thus,
+	// the security of that TLS connection is the same as the security of bootstrapping process that gave the node
+	// its client identity.
```

Comment:
that's a good argument on top of the existing mechanism.

is --kubelet-validate-node-name a new (tbd) flag for the apiserver?

i am interested if the flag should be enabled by default in kubeadm and if it will break existing setups. iiuc, it might.